### ISSUE TITLE
set.add(set, items, item, compare)

### ISSUE THREAD

**Issue by justinbmeyer** [canjs/can-set#2]
Add a method that adds an item to a given set in the right place.  Sorting should be taken into account.

**Comment by justinbmeyer** [comment]
Also, we could add a:

``` js
set.index(set, items, item, compare) //-> 1
```

**Comment by jandjorgensen** [comment]
Documented some of the formal set stuff in the test cases. Looks like intersection is the main problem. Only other issue I see is I don't think there's really a way to define a union of two sets in the object notation?

On this branch I have two failing intersection tests along with a lot of comments:
https://github.com/canjs/can-set/compare/master...jandjorgensen:intersection-bug

**Comment by justinbmeyer** [comment]
This should also work for `fixture` so it is able to sort the entire set when someone passes a param.  For example if I did:

```
var algebra = new set.Algebra( SORT_CONFIG );
var taskStore = fixture.store(100, function(i){
  return { name: i+"th task" }
},algebra)
```

where `algebra` has some configured sort rules, I should be able to do:

``` js
taskStore.getListData({sort: "name desc"})
```

and this will give me back tasks sorted correctly.  This can be provided with some function like:

``` js
function(sortPropValue, item1, item2){
  var parts = sortValue.split(" "),
      sortProp = parts[0],
      asc = parts[1] === "asc",
     item1Value = item1[sortProp],
     item2Value = item2[sortProp];

  if(!asc) {
    var temp = item1Value;
    item1Value = item2Value
    item2Value = item1Value;
  }

  if( item1Value < item2Value ) {
    return 1
  } else if(item1Value === item2Value ) {
    return 0
  } else {
    return -1;
  }
}
```

Another problem with this is ranged sets.  How do we know if the item actually belongs in the set? And how do we shift all the items around in every set?

For example, say I created an Algebra for a `todosConnection` like:

``` js
new Algebra( 
  set.comparators.rangeInclusive("start", "end"),
  SORT_CONFIG
)
```

Then I made requests like:

``` js
todosConnection.getList({start: 0, end: 1, sort: "name"})
//-> [{name: "a"}, {name: "b}]
todosConnection.getList({start: 3, end: 4, sort: "name"})
//-> [{name: "d"}, {name: "e"}]
```

Notice we don't know what's in the index=2 spot.  If there was a delete of todo `a`, do we request what is now in index=1 and index=4?

`rangeInclusive` is going to have to work closely with any `sort` logic.  Currently, can-connect's `real-time` calls `algebra.subset(item, set)` to check if an item is within a set.  

Lets say we created "clue".  But we didn't know if the old item at index=2 was "cap" or "cyclone".  How would `.subset()` work?  Without knowing what the current data looks like, it can't even hope to work:

```
algebra.subset({name: "clue"}, {start: 0, end: 1, sort: "name"}) //-> ?
```

However, with the current set data: we can know at least it's not in the first set:

```
algebra.subset({name: "clue"}, {start: 0, end: 1, sort: "name"}, items) //-> false
algebra.subset({name: "clue"}, {start: 3, end: 4, sort: "name"}, items) //-> true or false?
```

As for the 2nd set ... at least we know it's less than our old current value ... so we need to ask the server for more info.

tldr; mutations to unknown "holes" makes the sets need to ask the server for more information.

**Comment by justinbmeyer** [comment]
Perhaps a `set.inSet( item, items, set)`.  I think this would return `-1`, `0`, `1` or `true` or `false`. Numbers would be returned if you wanted to indicate a direction.  

I think the default behavior of real-time would be to run though each "ranged" set.  If the item is within a set, it will be added to the set at `set.index()`.  

If it's between sets, it will be added to the end of the first set that returns `1`.  And it will then shift every other set by one.

This means that if you don't have continuous sets, errors could happen.  But I don't think that's likely.  Instead you'll probably see inserts.

**Comment by justinbmeyer** [comment]
So SQL does the following clauses in order:

```
WHERE  
GROUPING
HAVING
ORDER
PAGINATE (limit/offset)
```

I think an algebra should assemble itself into different stages.  Say, for example, `rangeInclusive`, instead of returning a _POJO_ `compares` object, returned something like:

``` js
new set.clause.Paginate(compares)
```

Similarly, `sort` could return:

```
new set.clause.Sort(compares)
```

And enum / boolean:

```
new set.clause.Where(compares);
```

when an Algebra is assembled like:

``` js
new set.Algebra(
  comparators.sort('sort'),
  comparators.rangeInclusive('start','end'),
  comparators.boolean("complete")
);
```

internally, it will organize its arguments by their type.  Something like:

``` js
Algebra = function(){
  this.clauses = {where: [], sort: [], paginate: []}
  arguments.each(function(arg){
    // I'm assuming that each clause constructor function has a type property.
    // if it doesn't have a type, assume where
    clauses[arg.constructor.type].push(arg)
  })
}
```

With that in place, we can make an algebra provide different treatments to different clauses.  In reality, most clauses can be treated the same ... behaviors between clauses will be different.  I'll try to group behavior like:
- COMPARE methods like: `.subset` or `.equal`
- SET LOGIC methods like: `.difference` or `.intersection`
- SET DATA methods like: `.getSubset`

### ORDER clauses
-  COMPARE methods - The order clause is essentially ignored unless there is a LIMIT clause.  If there is a LIMIT clause, then the order clauses need to match.  
- SET LOGIC - If sets ORDERS match, that property should probably be in result. If not the same, not present.  
- SET DATA methods - `.getSubset` needs to be called.

**Comment by justinbmeyer** [comment]
Closed.  This was part of 0.5.0.
